org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ClasspathManagement.scala

```diff
@@ -493,5 +494,42 @@ trait ClasspathManagement extends HasLogger { self: ScalaProject =>
     errors.toSeq
   }
 
+  /** Make sure compiler plugins are binary compatible with the selected Scala installation.
+   *
+   *  Unlike regular dependencies, compiler plugins must be cross-compiled with the exact compiler
+   *  version used for building.
+   */
+  private def validateCompilerPlugins(): Seq[(Int, String, String)] = if (!isUsingCompatibilityMode()) Seq() else {
+    val errors = mutable.ListBuffer[(Int, String, String)]()
+
+    def paths(p: String) =
+      p.split(File.pathSeparator).toSeq.filterNot(_.isEmpty).distinct
+
+    def scanDir(d: String) = {
+      val entries = Option(new File(d).listFiles).getOrElse(Array.empty[File])
+      for (f <- entries if f.isFile)
+        yield f.getAbsolutePath
+    }
+
+    val plugins = paths(storage.getString("Xpluginsdir")).flatMap(scanDir) ++ paths(storage.getString("Xplugin"))
+    val installation = effectiveScalaInstallation()
```

Comment:
Compiler plugins can use internal methods, so they can't run inside the PC unless they match the exact version. The PC is always running the bundled Eclipse version (`platformInstallation`). `-Xsource` and compatibility mode are mostly about what the type-checker flags as errors: it will type-check "as-if" it was the 2.10 compiler, so any bugfixes in 2.11 that give different errors are behind the `-Xsource` flag.

However, the checks here are for for the build compiler. For building, we instantiate a `scala-compiler.jar` from the project Scala Installation, so if the compiler plugin matches that version, it should work.

Does this answer your question?